funsor/terms.py

```
@@ -1057,53 +1058,95 @@ class Stack(Funsor):
 
     def eager_reduce(self, op, reduced_vars):
-        components = self.components
+        parts = self.parts
         if self.name in reduced_vars:
             reduced_vars -= frozenset([self.name])
             if reduced_vars:
-                components = tuple(x.reduce(op, reduced_vars) for x in components)
-            return reduce(op, components)
-        components = tuple(x.reduce(op, reduced_vars) for x in components)
-        return Stack(components, self.name)
+                parts = tuple(x.reduce(op, reduced_vars) for x in parts)
+            return reduce(op, parts)
+        parts = tuple(x.reduce(op, reduced_vars) for x in parts)
+        return Stack(self.name, parts)
+
+
+class Cat(Funsor):
+    """
+    Concatenate funsors along an existing input dimension.
+
+    :param str name: The name of the input variable along which to concatenate.
+    :param tuple parts: A tuple of Funsors of homogenous output domain.
+    """
+    def __init__(self, name, parts):
+        assert isinstance(name, str)
+        assert isinstance(parts, tuple)
+        assert parts
+        assert all(name in x.inputs for x in parts)
+        assert len(set(x.output for x in parts)) == 1
+        output = parts[0].output
+        inputs = OrderedDict()
+        for x in parts:
+            inputs.update(x.inputs)
+        inputs[name] = bint(sum(x.inputs[name].size for x in parts))
+        super().__init__(inputs, output)
+        self.name = name
+        self.parts = parts
+
+
+@eager.register(Cat, str, tuple)
+def eager_cat(name, parts):
+    if len(parts) == 1:
+        return parts[0]
+    return eager_cat_homogeneous(name, *parts)
+
+
+@dispatch(str, Variadic[Funsor])
+def eager_cat_homogeneous(name, *parts):
+    return None  # defer to default implementation
+
+
+@dispatch(str, Variadic[Stack])
+def eager_cat_homogeneous(name, *parts):
+    parts = sum((p.parts for p in parts), ())
+    return Stack(name, parts)
```

This behavior isn't consistent with `Tensor` if `any(part.name != name for part in parts)` - maybe raise an error in that case?

---

Good catch! I'll fix shortly...

---

Fixed by deferring to default (lazy) interpretation.